### ISSUE TITLE
findAllAsync().asObservable() does not return loaded results sometimes.

### ISSUE THREAD

**Issue by rfrost77** [realm/realm-java#2243]
```
realm.where(GroupRealm.class).findAllAsync().asObservable()
                .filter(RealmResults::isLoaded).first();
```

It's working code for get groups by async with observable.

But sometimes, it does not return loaded results. After return placeholder results with size 0, it stops though it has some datas. In case of this, UI gets stuck because any subscribe event is never called.

How can I get guarantee of getting results from async observable methods?

**Comment by bclymer** [comment]
I've noticed this as well, using pretty much the exact same query. While debugging I ended up putting
```
// generate a guid
Log.d("Realm", guid + " - IsLoaded=" + query.isLoaded);
```
in the filter block, and probably 9/10 times I see two prints for each guid, with `isLoaded` as `false` and then `true`. But every once in a while I only see a guid printed once, with `isLoaded` as `false`.

**Comment by rfrost77** [comment]
@bclymer For me, it happens more frequently, almost 50% of cases.

**Comment by cmelchior** [comment]
@rfrost77 If you close the Realm instance it will stop emitting. Could this be the reason? If not, and you can reproduce it this frequently It would be great if you can create a small sample project we can investigate.

**Comment by rfrost77** [comment]
After getting advice from you 'If you close the Realm instance it will stop emitting.', I changed some realm close logic. And I think it may a little work(probability of not returning loaded realm results decreases).

I guess it is async timing issue. I write on realm by async method, get subscribe event from this method quitting, and read changed data on realm by also async method. This logic blocks are multiple at the same time, so realm opening and closing times are coming unpredictable. Below is by assumption of error logic.

Realm A opens, before closing Realm A, Realm B opens, and Realm A closes. But it effects as Realm B closing because these are opened from a same thread, therefore Realm B no longer emit item.

If my assumption has wrong, please point out and if it is right, how can I use safe realm opening and closing in multi thread?

**Comment by rfrost77** [comment]
@cmelchior In addition to I mentioned above, reading data by async after writing data on realm by async reads data which is before changed. Even though I absolutely read data after writing, It comes not changed data from realm by async. How can I read 100% changed data after async writing?

**Comment by nhachicha** [comment]
@rfrost77 
not all writes triggers a `RealmChangeListener` (which will end up emitting an observable) 

if you query on `GroupRealm` then you write on `Foo` this will not trigger the listener for your async query on `GroupRealm`.

It would be helpful to provide a sample code/test (or a pseudo code) to follow your use case.

**Comment by cmelchior** [comment]
Hi @rfrost77 Did you figure out a way to get it working?

**Comment by cmelchior** [comment]
Closing due to no response

**Comment by ganfra** [comment]
Hello @cmelchior , I have the same issue than described at first.
Sometime, the findAllAsync.asObservable return only one unload result. The realm instance is not closed during the process ...

`public static class TestFragment extends Fragment { `

        public static TestFragment newInstance() {
            Bundle args = new Bundle();
            TestFragment fragment = new TestFragment();
            fragment.setArguments(args);
            return fragment;
        }

        private Realm mRealm;
        private Subscription mTestSubscription;

        @Override
        public void onActivityCreated(@Nullable Bundle savedInstanceState) {
            super.onActivityCreated(savedInstanceState);
            mRealm = Realm.getDefaultInstance();
        }

        @Override
        public void onStart() {
            super.onStart();
            mTestSubscription = getQuerySubscription();
        }


        @Override
        public void onStop() {
            mTestSubscription.unsubscribe();
            super.onStop();
        }

        @Override
        public void onDestroy() {
            mRealm.close();
            super.onDestroy();
        }

        private Subscription getQuerySubscription() {
            return mRealm.where(ProfileView.class).findAllAsync().asObservable()
                    .filter(new Func1<RealmResults<ProfileView>, Boolean>() {
                        @Override
                        public Boolean call(RealmResults<ProfileView> profileViews) {
                            return profileViews.isLoaded();
                        }
                    })
                    .subscribe(new Subscriber<RealmResults<fr.ganfra.rxtest.RxRealmTest.ProfileView>>() {
                        @Override
                        public void onCompleted() {
                            // DO SOMETHING
                        }

                        @Override
                        public void onError(Throwable e) {
                            // DO SOMETHING
                        }

                        @Override
                        public void onNext(RealmResults<fr.ganfra.rxtest.RxRealmTest.ProfileView> profileViews) {
                            // DO SOMETHING
                        }
                    });
        }

    }


    public static class ProfileView extends RealmObject {

        private long dateOfBirth;
        private String deliveryMode;
        private String email;
        private String firstName;
        private String fixePhoneNumber;
        private String gender;
        @PrimaryKey
        @SerializedName("id")
        private String identifier;
        private boolean isDeliveryAdressDifferent;
        private boolean isReceivingNewsletter;
        private boolean isReceivingPartnersNewsletter;
        private boolean isReceivingSms;
        private boolean ladNotAllowed;
        private String lastName;
        private String login;
        private String mobilePhoneNumber;

        public long getDateOfBirth() {
            return dateOfBirth;
        }

        public void setDateOfBirth(long dateOfBirth) {
            this.dateOfBirth = dateOfBirth;
        }

        public String getDeliveryMode() {
            return deliveryMode;
        }

        public void setDeliveryMode(String deliveryMode) {
            this.deliveryMode = deliveryMode;
        }

        public String getEmail() {
            return email;
        }

        public void setEmail(String email) {
            this.email = email;
        }

        public String getFirstName() {
            return firstName;
        }

        public void setFirstName(String firstName) {
            this.firstName = firstName;
        }

        public String getFixePhoneNumber() {
            return fixePhoneNumber;
        }

        public void setFixePhoneNumber(String fixePhoneNumber) {
            this.fixePhoneNumber = fixePhoneNumber;
        }

        public String getGender() {
            return gender;
        }

        public void setGender(String gender) {
            this.gender = gender;
        }

        public String getIdentifier() {
            return identifier;
        }

        public void setIdentifier(String identifier) {
            this.identifier = identifier;
        }

        public boolean isDeliveryAdressDifferent() {
            return isDeliveryAdressDifferent;
        }

        public void setIsDeliveryAdressDifferent(boolean isDeliveryAdressDifferent) {
            this.isDeliveryAdressDifferent = isDeliveryAdressDifferent;
        }

        public boolean isReceivingNewsletter() {
            return isReceivingNewsletter;
        }

        public void setIsReceivingNewsletter(boolean isReceivingNewsletter) {
            this.isReceivingNewsletter = isReceivingNewsletter;
        }

        public boolean isReceivingPartnersNewsletter() {
            return isReceivingPartnersNewsletter;
        }

        public void setIsReceivingPartnersNewsletter(boolean isReceivingPartnersNewsletter) {
            this.isReceivingPartnersNewsletter = isReceivingPartnersNewsletter;
        }

        public boolean isReceivingSms() {
            return isReceivingSms;
        }

        public void setIsReceivingSms(boolean isReceivingSms) {
            this.isReceivingSms = isReceivingSms;
        }

        public boolean isLadNotAllowed() {
            return ladNotAllowed;
        }

        public void setLadNotAllowed(boolean ladNotAllowed) {
            this.ladNotAllowed = ladNotAllowed;
        }

        public String getLastName() {
            return lastName;
        }

        public void setLastName(String lastName) {
            this.lastName = lastName;
        }

        public String getLogin() {
            return login;
        }

        public void setLogin(String login) {
            this.login = login;
        }

        public String getMobilePhoneNumber() {
            return mobilePhoneNumber;
        }

        public void setMobilePhoneNumber(String mobilePhoneNumber) {
            this.mobilePhoneNumber = mobilePhoneNumber;
        }
    }

**Comment by cmelchior** [comment]
Hi @ganfra 

Unsubscribing will also stop any callback. Can you verify that `onStop` is not being called when you see this?

**Comment by ganfra** [comment]
@cmelchior, onStop is not called. However, the behavior seems to disappear in debug mode with breakpoints, it's really weird

**Comment by cmelchior** [comment]
Can you reproduce it consistently? If possible, can you extract the behaviour in a sample project we can investigate?

**Comment by ganfra** [comment]
I can't reproduce every time, so it's not very convenient to debug ...

**Comment by ganfra** [comment]
I just implements RxObservableFactory, using the same as yours, to put some breakpoints.
And it sounds like RealmChangeListener is not called when results get loaded

**Comment by ganfra** [comment]
Any idea @cmelchior  ? I have replaced all the getAllAsync call with getAll, and the problem doen't appear anymore ...

**Comment by joluet** [comment]
@cmelchior I'm observing the same issue. `findAllAsync()` sometimes doesn't return any results. Whereas `findAll()` works reliably. I'm using version `2.2.1` with `syncEnabled = true`. 

I'm curious what's the drawback of using `findAll()` in a query like this?:
```
final Realm realm = Realm.getDefaultInstance();
return realm.where(Trip.class)
                .equalTo("uid", tripId)
                .findAll()
                .asObservable()
                .map(realm::copyFromRealm)
                .doOnUnsubscribe(realm::close);
```
Will this block the UI thread if there are many `Trip` objects in the Realm? What's the advantage of using `findAllAsync()` as described in the docs?

**Comment by cmelchior** [comment]
@joluet If you have an example app where you can reproduce notifications being missed, I would be extremely interested in it.

`findAllAsync` will do the query on a background thread and only return the result once the query is done. `findAll()` will run the query on the UI thread. In your case I expect it is the `copyFromRealm()` call that is expensive though which is being executed on the UI thread.

**Comment by Zhuinden** [comment]
@joluet 

1.) Typically `realm.copyFromRealm()` is unnecessary, and means you're misusing something

2.) `findAllAsync()` must be filtered with `isLoaded()`

    return realm.where(Trip.class)
                .equalTo("uid", tripId)
                .findAll()
                .asObservable()
                .filter(RealmResults::isLoaded);

**Comment by joluet** [comment]
Actually, I only see that issue when the app is connected to the Realm Object Server. So it seems to be related to the sync feature.

**Comment by joluet** [comment]
@cmelchior The issue that I'm seeing is definitely on happening in sync mode when connected to the Realm Object Server. Also, I was able to modify the RealmTasks example app to reproduce the issue. I created a new issue (#3938) as it's probably not related to this one.
